### PR TITLE
Add polling instance status after async operations

### PR DESCRIPTION
**Pull request by leonwanghui**
This PR is proposed for solving #46 by adding polling instance status method after async (provision, update and deprovision) operations.

### CODE REVIEW

File: 2.13/tests/test/test.js [hunk 1]
```diff
@@ -473,10 +445,52 @@ function testDeprovision(instanceId, queryStrings, isAsync){
...
+                    if (message != "") return done(new Error(message));
+                    lastOperationState = res.body.state;
+                })
+            setTimeout(callback(null, count), config.pollingInterval*1000);
+        },
+        function(err, n) {
+            if (err) {
+                return done(new Error("Polling last operation error!"))
+            } else if (lastOperationState == "failed") {
+                return done(new Error("Polling last operation stateus failed!"));
```
Comment (zhongyi-zhang) [inline 2.13/tests/test/test.js]: "stateus" typo?

Comment (leonwanghui) [reply 2.13/tests/test/test.js]: Already fix it, thanks!

File: 2.13/tests/test/test.js [hunk 2]
```diff
@@ -330,9 +295,21 @@ function testUpdate(instanceId, validBody, isAsync){
...
-                testLastOperationStatus(res.body, done)
+                else done();
             })
         });
+
+        if (isAsync) {
+            // TODO: the query string should contain 'operation' for the last operation. FYI: https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md#parameters.
+            describe('UPDATE - poll', function() {
+                testAPIVersionHeader('/v2/service_instances/' + instanceId + '/last_operation', 'GET');
+                testAuthentication('/v2/service_instances/' + instanceId + '/last_operation', 'GET');
```
Comment (norshtein) [inline 2.13/tests/test/test.js]: Not need to test version header and authentication again here. This path is already tested in `describe('PROVISION - poll')`

Comment (norshtein) [reply 2.13/tests/test/test.js]: Please help remove the duplicate test, thanks!

**Comment by norshtein** [comment]
LGTM

**Comment by leonwanghui** [comment]
Hi @norshtein @zhongyi-zhang , when I tested this patch with some real services, some errors occur as below:
```shell
PROVISION - new
should accept a valid provision request9476ms
PROVISION - poll
should reject requests without X-Broker-API-Version header with 412
should reject unauthorized requests with 401
should reject bad credentials with 401
should return succeeded operation status after provision
should return succeeded operation status after provision
Error: done() called multiple times
    at pollInstanceLastOperationStatus (test/test.js:502:5)
    at Context.<anonymous> (test/test.js:238:21)
```

Do you know how to fix it? Thanks!

File: 2.13/tests/test/test.js [hunk 3]
```diff
@@ -473,10 +447,55 @@ function testDeprovision(instanceId, queryStrings, isAsync){
...
+            if (err) {
+                return done(new Error("Polling last operation error!"))
+            } else if (lastOperationState == "failed") {
+                return done(new Error("Polling last operation status failed!"));
+            } else {
+                return done();
+            }
+        }
+    );
+    done();
```
Comment (zhongyi-zhang) [inline 2.13/tests/test/test.js]: The `return` above scopes at its current function. So, here `done()` is called another time after `return done` above.

File: 2.13/tests/test/test.js [hunk 4]
```diff
@@ -473,10 +447,55 @@ function testDeprovision(instanceId, queryStrings, isAsync){
...
+            preparedRequest()
+                .get('/v2/service_instances/' + instanceId + '/last_operation')
+                .set('X-Broker-API-Version', apiVersion)
+                .auth(config.user, config.password)
+                .expect(200)
+                .expect('Content-Type', /json/)
+                .end(function(err, res){
+                    if (err) return done(err);
+                    var message = validateJsonSchema(res.body, lastOperationSchema);
+                    if (message != "") return done(new Error(message));
```
Comment (zhongyi-zhang) [inline 2.13/tests/test/test.js]: Sorry I didn't review it carefully... Also, here and line 479 `done` should not be called. Just `callback(err)` and leave the `done` calling to the result handling function below.

Comment (leonwanghui) [reply 2.13/tests/test/test.js]: Ok, thanks!

File: 2.13/tests/test/test.js [hunk 5]
```diff
@@ -473,10 +447,57 @@ function testDeprovision(instanceId, queryStrings, isAsync){
                     done();
                 })
             });
+
+            if (isAsync) {
+                // TODO: the query string should contain 'operation' for the last operation. FYI: https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md#parameters.
+                describe('DEPROVISION - poll', function() {
+                    it ('should return succeeded operation status after deprovision', function(done){
+                        pollInstanceLastOperationStatus(instanceId, done);
```
Comment (zhongyi-zhang) [inline 2.13/tests/test/test.js]: @leonwanghui as you can see, the default timeout for each case is 2000ms in mocha, have a try `this.timeout(timeOutInMs)`. Ref: https://mochajs.org/#timeouts

Comment (zhongyi-zhang) [reply 2.13/tests/test/test.js]: And it should be also applied in provision-poll.

File: 2.13/tests/test/test.js [hunk 6]
```diff
@@ -473,10 +447,57 @@ function testDeprovision(instanceId, queryStrings, isAsync){
...
+                    var message = validateJsonSchema(res.body, lastOperationSchema);
+                    if (message != "") return callback(new Error(message), count);
+                    lastOperationState = res.body.state;
+                })
+            setTimeout(function() {
+                callback(null, count);
+            }, config.pollingInterval*1000);
+        },
+        function(err, n) {
+            console.log(n + 'th polling last operation...');
```
Comment (zhongyi-zhang) [inline 2.13/tests/test/test.js]: Should it be moved to the first function in this `async.whilst`? It only reaches here after the loop ends or it catches an error.

File: 2.13/tests/test/test.js [hunk 7]
```diff
@@ -465,23 +469,26 @@ function pollInstanceLastOperationStatus(instanceId, done) {
...
                 .set('X-Broker-API-Version', apiVersion)
                 .auth(config.user, config.password)
                 .expect(200)
                 .expect('Content-Type', /json/)
                 .end(function(err, res){
-                    if (err) return done(err);
+                    if (err) callback(err, count);
                     var message = validateJsonSchema(res.body, lastOperationSchema);
-                    if (message != "") return done(new Error(message));
+                    if (message != "") return callback(new Error(message), count);
```
Comment (zhongyi-zhang) [inline 2.13/tests/test/test.js]: When calling an error back, you can only call the error back as suffix function variables can be absent in Nodejs. And it is a habit in Nodejs, always putting the error in the first place and always checking if the error exists first in a callback.

**Comment by zhongyi-zhang** [comment]
LGTM.

[Approved with PullApprove](https://pullapprove.com/openservicebrokerapi/osb-checker/pull-request/52/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=openservicebrokerapi/osb-checker)

**Comment by norshtein** [comment]
LGTM

**Pull request merged**